tests/models/test_hparams.py

```diff
@@ -642,3 +648,46 @@ def test_model_with_fsspec_as_parameter(tmpdir):
     )
     trainer.fit(model)
     trainer.test()
+
+
+@pytest.mark.skipif(not HYDRA_EXPERIMENTAL_AVAILABLE, reason="Hydra experimental is not available")
+def test_model_save_hyper_parameters_interpolation_with_hydra(tmpdir):
+    """
+    This test relies on configuration saved under tests/models/conf/config.yaml
+    """
+
+    class TestHydraModel(BoringModel):
+
+        def __init__(self, args_0, args_1, args_2, kwarg_1=None):
+            self.save_hyperparameters()
+            self.test_hparams()
+            config_file = f"{tmpdir}/hparams.yaml"
+            save_hparams_to_yaml(config_file, self.hparams)
+            self.hparams = load_hparams_from_yaml(config_file)
+            self.test_hparams()
+            super().__init__()
+
+        def test_hparams(self):
+            assert self.hparams.args_0.log == "Something"
+            assert self.hparams.args_1['cfg'].log == "Something"
+            assert self.hparams.args_2[0].log == "Something"
+            assert self.hparams.kwarg_1['cfg'][0].log == "Something"
+
+    with initialize(config_path="conf"):
+        args_0 = compose(config_name="config")
+        args_1 = {"cfg": compose(config_name="config")}
+        args_2 = [compose(config_name="config")]
+        kwarg_1 = {"cfg": [compose(config_name="config")]}
+        model = TestHydraModel(args_0, args_1, args_2, kwarg_1=kwarg_1)
```

Comment:
Is this trying to emulate the following?
```
cfg = compose(config_name="config")
model = TestHydraModel(**cfg) 
```

Reply:
The function call itself does this:

```python
class MyModel(pl.LightningModule):
    def __init__(self,
                 cfg_1: DictConfig,
                 cfg_2: Dictconfig):
        self.save_hyperparameters() # Autosaves your parameters for loading your class in the future

# after training your model
MyModel.load_from_checkpoint('model.ckpt') # load's from checkpoint, uses the saved input args 
```

In the test he's emulating the above I think!

Reply:
I think the logic flow that is currently implemented is:

- Loop through the hparams dictionary
- If DictConfig, resolve interpolation and create new dict container
- After the loop, save all the resolved DictConfigs

There is some weirdness however. I'm hoping we can do something a bit cleaner:

- If OmegaConf is available, we rely on OmegaConf entirely to save the hparams dictionary without resolving anything. @omry we should expect OmegaConf.save to work with primitive types/anything that can be pickle-able right?
- If not, default save functionality

Reply:
> If OmegaConf is available, we rely on OmegaConf entirely to save the hparams dictionary without resolving anything. @omry we should expect OmegaConf.save to work with primitive types/anything that can be pickle-able right? 

No, OmegaConf does not support arbitrary objects in the config.

> How about having two different DictConfigs within the dictionary structure?

That should be fine.